api/src/controllers/changes.js

```diff
@@ -1,578 +1,396 @@
-  return subjectIds.indexOf(submitter) === -1;
 };
 
-const getValidatedDocIds = (contactId, facilityId, subjectIds, userCtx, callback) => {
-  const startTime = startTimer();
-  db.medic.view('medic', 'docs_by_replication_key', { keys: subjectIds }, function(err, viewResult) {
-    endTimer('bindValidatedDocIds().docs_by_replication_key', startTime);
-    if (err) {
-      return callback(err);
-    }
-    const validatedIds = viewResult.rows.reduce(function(ids, row) {
-      if (!isSensitive(contactId, facilityId, subjectIds, row.key, row.value.submitter)) {
-        ids.push(row.id);
-      }
-      return ids;
-    }, [ '_design/medic-client', 'org.couchdb.user:' + userCtx.name ]);
-    endTimer('bindValidatedDocIds().before-callback', startTime);
-    callback(null, validatedIds);
-  });
-};
-
-var bindRequestedIds = function(feed, callback) {
-  var ids = [];
-  if (feed.req.body && feed.req.body.doc_ids) {
-    // POST request
-    ids = feed.req.body.doc_ids;
-  } else if (feed.req.query && feed.req.query.doc_ids) {
-    // GET request
-    try {
-      ids = JSON.parse(feed.req.query.doc_ids);
-    } catch(e) {
-      return callback({ code: 400, message: 'Invalid doc_ids param' });
-    }
-  }
-  feed.requestedIds = ids;
-  return callback();
+const isLongpoll = req => {
+  return req.query && ['longpoll', 'continuous', 'eventsource'].indexOf(req.query.feed) !== -1;
 };
 
-var defibrillator = function(feed) {
+const defibrillator = feed => {
   if (feed.req.query && feed.req.query.heartbeat) {
-    const heartbeatInterval = heartbeatFilter(feed.req.query.heartbeat);
-    feed.heartbeat = setInterval(function() {
-      feed.res.write('\n');
-    }, heartbeatInterval);
+    const heartbeat = heartbeatFilter(feed.req.query.heartbeat);
+    feed.heartbeat = setInterval(() => writeDownstream(feed, '\n'), heartbeat);
   }
 };
 
-var prepareResponse = function(feed, changes) {
-  if (feed.res.finished) {
-    // Don't write to the response if it has already ended. The change
-    // will be picked up in the subsequent changes request.
-    return;
+const addTimeout = feed => {
+  if (feed.timeout) {
+    clearTimeout(feed.timeout);
   }
-  // filter out records the user isn't allowed to see
-  changes.results = changes.results.filter(function(change) {
-    return change.deleted || _.contains(feed.validatedIds, change.id);
-  });
-  feed.res.write(JSON.stringify(changes));
-};
-
-function abortUpstreamBatches(feed, reason, logMessage) {
-  const uBatchCount = Object.keys(feed.upstreamBatches).length;
-  feed.log(`[abortUpstreamBatches():${reason}]`, `Cancelling ${uBatchCount} batches...`);
 
-  Object.keys(feed.upstreamBatches).forEach(id => {
-    feed.log(`[abortUpstreamBatches():${reason}]`, logMessage, id);
+  const timeout = (feed.req.query && feed.req.query.timeout) ?
+    Math.min(DEFAULT_TIMEOUT, feed.req.query.timeout) : DEFAULT_TIMEOUT;
+  feed.timeout = setTimeout(() => endFeed(feed), timeout);
```

Comment:
This is a new feature, right? Why did you decide the feed should timeout every 60s? The 2.x version should run forever unless a change is received or the connection drops. I think this is the desired behaviour.